### PR TITLE
Trigger build for projects configured with `BitbucketSCM`

### PR DESCRIPTION
**Pull request by justinskariah**
* Trigger build for projects configured with `BitbucketSCM` SCM when refs changed webhook is received
* `BitbucketWebhookConsumer` code has been refactored to make it more readable.

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 1]
```diff
@@ -26,90 +28,144 @@
...
+        }
+    }
+
+    private static class RefChangedDetails {
+
+        private final Set<String> cloneLinks;
+        private final BitbucketRepository repository;
+
+        public RefChangedDetails(RefsChangedWebhookEvent event) {
+            this.cloneLinks = cloneLinks(event);
```
Comment (khughes-atlassian) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java]: Unnecessary `this` /nit

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 2]
```diff
@@ -26,90 +28,144 @@
...
+
+    private static Collection<? extends SCM> getScms(ParameterizedJobMixIn.ParameterizedJob job) {
+        SCMTriggerItem triggerItem = SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(job);
+        if (triggerItem != null) {
+            return triggerItem.getSCMs();
+        }
+        return Collections.emptySet();
+    }
+
+    private static class TriggerDetails {
```
Comment (khughes-atlassian) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java]: This class and `RefChangedDetails` can both be final and have private constructors

**Comment by justinskariah** [comment]
@khughes-atlassian
> * When `RefsChangedWebhookEvent.getActor` returns `null` (currently this results in an NPE I think)
`RefsChangedWebhookEvent.getActor` cannot be `null` for a ref changed event, it will be for mirror synchronised. So when we add support for mirror synchronized event it should be covered. It will not cause NPE because `BitbucketWebhookTriggerRequest.builder().actor()` handles `null` value.
> * When `triggerFrom` returns `null`
I thought `nullProject` would handle it, but it didn't because it had a trigger. I have created an ignored project to cover that case.

> * When `toTriggerDetails` returns `empty`
Is covered in `testRefsChangedShouldNotTriggerBitbucketSCMIfRepositoryDoesNotMatch`
> * When `getScms` returns `empty`
Good catch, it's now coverd by `nullProject`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 3]
```diff
@@ -26,90 +28,144 @@
...
+                    });
         }
     }
 
     void process(MirrorSynchronizedWebhookEvent event) {
         LOGGER.debug("Received mirror synchronized event" + event);
     }
 
-    private static Set<String> getCloneLinks(RefsChangedWebhookEvent event) {
+    private static Set<String> cloneLinks(RefsChangedWebhookEvent event) {
```
[the remark this answers is not in the document]
Comment (justinskariah) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java]: I wanted to keep `RefChangedDetails` as a simple POJO class

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java [hunk 4]
```diff
@@ -60,8 +60,8 @@
     private final List<GitSCMExtension> extensions;
     private final String gitTool;
     private final String id;
-    private final List<BitbucketSCMRepository>
-            scmRepositories; // this is to enable us to support future multiple repositories and
+    private final List<BitbucketSCMRepository> repositories; // this is to enable us to support future multiple repositories
```
[the remark this answers is not in the document]
Comment (justinskariah) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java]: There weren't any `BitbucketRepository` in `BitbucketSCM` and considering there is only one type of repo I renamed it to `repositories`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 5]
```diff
@@ -26,90 +28,144 @@
...
+
+    private static boolean hasMatchingRepository(RefChangedDetails refChangedDetails,
+                                                 GitSCM scm) {
+        return scm.getRepositories().stream()
+                .anyMatch(scmRepo -> matchingRepo(refChangedDetails.getCloneLinks(), scmRepo));
+    }
+
+    private static boolean matchingRepo(BitbucketRepository repository, BitbucketSCMRepository scmRepo) {
+        return scmRepo.getProjectKey().equals(repository.getProject().getKey())
+                && scmRepo.getRepositorySlug().equals(repository.getSlug());
```
[the remark this answers is not in the document]
Comment (justinskariah) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java]: Good catch, let me see what can be done. The issue is dealing with mirror synchronised event, which doesn't have `links`

[the remark this answers is not in the document]
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java]: I'm happy for us to merge this, and solve the above in a later release. That way we can keep the PRs small. If you agree can you open a ticket (if you don't agree and want to fix it as part of this PR I'm happy with that as well).

[the remark this answers is not in the document]
Comment (justinskariah) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java]: I have fixed the issue by comparing the base url with the self url

**Pull request merged**